kore/src/Kore/Step.hs

```diff
@@ -163,13 +169,40 @@ anyRewrite
 anyRewrite rewrites =
     Strategy.any (rewriteStep <$> rewrites)
 
+priorityAllStrategy
+    :: ToRulePattern rewrite
+    => [rewrite]
+    -> Strategy (Prim rewrite)
+priorityAllStrategy rewrites =
+    Strategy.any (fmap allRewrites priorityGroups)
+  where
+    priorityGroups =
+        groupSortOn
+            (getPriorityOfRule . toRulePattern)
+            rewrites
+
+priorityAnyStrategy
+    :: ToRulePattern rewrite
+    => [rewrite]
+    -> Strategy (Prim rewrite)
+priorityAnyStrategy rewrites =
+    anyRewrite sortedRewrites
+  where
+    sortedRewrites =
+        sortOn
+            (getPriorityOfRule . toRulePattern)
+            rewrites
+
 {- | Heat the configuration, apply a normal rewrite, and cool the result.
  -}
 -- TODO (thomas.tuegel): This strategy is not right because heating/cooling
 -- rules must have side conditions if encoded as \rewrites, or they must be
 -- \equals rules, which are not handled by this strategy.
 heatingCooling
-    :: (forall rewrite. [rewrite] -> Strategy (Prim rewrite))
+    ::  ( forall rewrite
+         .  ToRulePattern rewrite
```

Comment:
Why this new constraint?

Reply:
If I remember correctly, I think this was needed to be able to combine heatingCooling with `priorityAllStrategy`, which has this additional constraint.